File: Depth-First Search/Binary tree/Searching/Medium/LET_no98_Validate Binary Search Tree(iterative).py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        stack = []
        while root:
            stack.append(root)
            root = root.left
            
        prev = None
        while stack:
            cur = stack.pop()
            if prev and prev.val >= cur.val:
                return False
            prev = cur
            
            if cur.right:
                cur = cur.right
                while cur:
                    stack.append(cur)
                    cur = cur.left
        
        return True

File: Depth-First Search/Binary tree/Searching/Medium/LET_no98_Validate Binary Search Tree(iterative).cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        vector<TreeNode*> stack;
        while (root) {
            stack.push_back(root);
            root = root->left;
        }

        TreeNode *pre = nullptr, *cur;
        while (!stack.empty()) {
            cur = stack.back();
            stack.pop_back();
            if (pre && pre->val >= cur->val)
                return false;

            pre = cur;
            if (cur->right) {
                cur = cur->right;
                while (cur) {
                    stack.push_back(cur);
                    cur = cur->left;
                }
            }
        }

        return true;
    }
};